### PR TITLE
Fix/messaging

### LINKED ISSUE #178: Fix the Messaging UI and the Messaging button in the Lesson Info

### PR DESCRIPTION
**Pull request by Philipp0206**
## 🎯 PR Description:
This pull request solves some isues with the messaging feature. It also updates the previous messaging opportunity in confirmed lesson that went using an external messaging app to our own messaging feature. I want to thank @alexs-rufer for the help he gave me. No screenshots are included as there is nothing new in the UI.

## 📁 Files Changed:

- 📂 ConfirmedLessonTest.kt
Added a mocking of the chatviewmodel as it is needed in as argument.

- 📂 OverViewTest.kt
Updated the name, it has changed in the viewmodel so updated the name here too.

- 📂 MainActivity.kt
Adjusted the arguments of some functions as I changed their signature.

- 📂 ChatViewModel.kt
Changed the name of connect to conncectUser and slightly changed the implementation of the function. Slightly improved the ensureChannelExists and changed the name into createOrGetChannel.

- 📂 ConfirmedLesson.kt
Removed the old version of messaging and updated it by our own method in the messagebutton.

- 📂 Channel.kt
Improved functions for better usability used in confirmedLesson and chat. Improved nomenclature of the screen for better understanding of the user.

- 📂 Chat.kt
Rearranged the code so that the bug of the goBack button is solved. And removed bad and useless comments.

- 📂 Overview.kt
Updated the name since it has been changed in the viewmodel.

Close #178

### CODE REVIEW

File: app/src/androidTest/java/com/github/se/project/ui/lesson/ConfirmedLessonTest.kt
Comment (A-M-00) [inline app/src/androidTest/java/com/github/se/project/ui/lesson/ConfirmedLessonTest.kt:1]: Good to update the old tests when adding a new feature

File: app/src/main/java/com/github/se/project/MainActivity.kt
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/MainActivity.kt:1]: This is a lot of lines but seems to actually be a minor change

File: app/src/main/java/com/github/se/project/model/chat/ChatViewModel.kt [hunk 1]
```diff
@@ -3,36 +3,30 @@ package com.github.se.project.model.chat
...
-  val currentChannelID = currentChannelID_.asStateFlow() // Exposes an immutable flow for observers
+class ChatViewModel(private val chatClient: ChatClient) : ViewModel() {
 
-  // Tracks whether the user is already connected to avoid duplicate connections
-  private var connected = false
+  private val _currentChannelId = MutableStateFlow<String?>(null)
+  val currentChannelId = _currentChannelId.asStateFlow()
 
-  // Represents the initialization state of the ChatClient (e.g., not initialized, initializing, or
-  // complete)
```
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/model/chat/ChatViewModel.kt:27]: I don't mind the code changes, but why remove the comments?

File: app/src/main/java/com/github/se/project/model/chat/ChatViewModel.kt [hunk 2]
```diff
@@ -43,52 +37,103 @@ open class ChatViewModel(val chatClient: ChatClient) : ViewModel() {
...
+        )
+
+    val token = chatClient.devToken(user.id)
+
+    chatClient.connectUser(user, token).enqueue { result ->
+      if (result.isSuccess) {
+        isConnected = true
+        Log.d("ChatViewModel", "User connected successfully.")
+      } else {
+        Log.e("ChatViewModel", "Error connecting user: ${result.errorOrNull()}")
```
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/model/chat/ChatViewModel.kt:60]: Ideally these logs should be removed

File: app/src/main/java/com/github/se/project/model/chat/ChatViewModel.kt [hunk 3]
```diff
@@ -43,52 +37,103 @@ open class ChatViewModel(val chatClient: ChatClient) : ViewModel() {
...
+  ) {
+    val tutorId = lesson.tutorUid
+    val studentId = lesson.studentUid
+
+    // Generate a consistent channel ID
+    val channelId = generateChannelId(tutorId[0], studentId)
+
+    val channelMembers = listOf(tutorId[0], studentId)
+
+    var extraData = mapOf("name" to "$otherUsername")
```
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/model/chat/ChatViewModel.kt:91]: Nice feature, makes it more readable when having multiple chats

File: app/src/main/java/com/github/se/project/ui/lesson/ConfirmedLesson.kt [hunk 4]
```diff
@@ -231,27 +230,22 @@ private fun LessonDetailsCard(
...
-                  context, context.getString(R.string.inform_user_offline), Toast.LENGTH_SHORT)
-              .show()
-        }
+        chatViewModel.createOrGetChannel(
+            lesson,
+            { channel ->
+              chatViewModel.setCurrentChannelId(channel.cid)
+              navigationActions.navigateTo(Screen.CHAT)
+            },
+            lesson.title)
```
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/ui/lesson/ConfirmedLesson.kt:248]: This new version is much more readable. Why was the isConnected check removed though?

**Review by A-M-00 — APPROVED** [review]
Overall, decent PR, that enables new features and fixes previous issues. Should ideally get cleaned up a little though

File: app/src/androidTest/java/com/github/se/project/ui/lesson/ConfirmedLessonTest.kt
Comment (A-M-00) [inline app/src/androidTest/java/com/github/se/project/ui/lesson/ConfirmedLessonTest.kt:1]: Commenting out whole tests isn't ideal, Lea created a section in the wiki to stash tests if you need to delete them

File: app/src/androidTest/java/com/github/se/project/ui/overview/OverViewTest.kt
Comment (A-M-00) [inline app/src/androidTest/java/com/github/se/project/ui/overview/OverViewTest.kt:1]: These tests were definitely not all broken before. Are you sure it isn't due to your new features?

**Review by A-M-00 — APPROVED** [review]
Already reviewed it before, and not much changed. When adding new features though, you really should make sure they don't break previously-functioning tests.

File: app/src/androidTest/java/com/github/se/project/ui/overview/OverViewTest.kt
Comment (Philipp0206) [reply app/src/androidTest/java/com/github/se/project/ui/overview/OverViewTest.kt:1]: Yeah I will do that next week. I'm not sure to understand where the problems comes from for now. I'll look into it asap.

**Pull request merged**